### PR TITLE
Integrate vendor legal basis

### PR DESCRIPTION
**Pull request by allisonking**
Closes https://github.com/ethyca/fides/issues/4209
Partially addresses https://github.com/ethyca/fides/issues/4210

### Description Of Changes

Whew ok this is kind of a big one! The backend API had significant changes in order to add legal basis as another dimension. Purposes, Vendors, and Systems all have this new dimension. Whereas before, our experience object looked like

```js
{
  ...
  tcf_purposes: [...],
  tcf_vendors: [...],
  tcf_systems: [...],
}
```

it now looks like

```js
{
  ...
  tcf_consent_purposes: [...],
  tcf_legitimate_interests_purposes: [...],
  tcf_consent_vendors: [...],
  tcf_legitimate_interests_vendors: [...],
  tcf_consent_systems: [...],
  tcf_legitimate_interests_systems: [...],
  
  tcf_vendor_relationships: [...],
  tcf_system_relationships: [...]
}
```

In other words, we have a lot more fields to work with! the `*relationships` fields are to deal with fields that the vendor/system fields have in common, in order not to repeat too much in the payload. More context about that [here](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2800451585/2023-09-30+Saving+Backend+Legal+Basis+Dimension+alongside+Purposes+and+Vendors#Other-ideas-explored).

At the same time, the FE had gone ahead and implemented double vendor toggles, saving only to the TC string. Now that the backend can save it, we can get rid of some tech debt code 💥 

### Code Changes

* [x] Update TS types in both admin-ui and fides-js
* [x] Add EEA to human readable regions
* [x] Delete various files we don't need anymore (`CookieTable` (never used) and `LegalBasisDropdown` (no longer needed))
* [x] Refactor vendors/systems throughout to handle legal basis coming from the backend
* [x] Lots of refactoring to make purposes also handle a legal basis dimension 
* [x] Update cypress tests (this is a big diff because the payload data changed so much)

### Steps to Confirm

* [Set up your TCF environment](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2795438113/Setting+up+TCF)
* Add various purposes of either purpose or legitimate interest types. Add them to different vendors
* Load the privacy center and visit `/fides-js-demo.html?geolocation=fi-18`. You should see the prefetched experience in the HTML. this should have all the new fields
* Make sure everything still works
* The one thing that won't work is you won't be able to change Legitimate Interest Purposes individually yet (opting in or out of all will work). We need to add the second toggle to that tab to do so, but I split that out into a [separate ticket](https://github.com/ethyca/fides/issues/4210)

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [x] Issue Requirements are Met
* [x] Update `CHANGELOG.md`

### CODE REVIEW

File: clients/fides-js/src/components/tcf/InitialLayer.tsx [hunk 1]
```diff
@@ -7,46 +7,67 @@ import {
...
-      experience.tcf_purposes ? experience.tcf_purposes.map((p) => p.id) : [],
-    [experience.tcf_purposes]
-  );
+  const {
+    tcf_consent_purposes: consentPurposes = [],
+    tcf_legitimate_interests_purposes: legintPurposes = [],
+    tcf_special_features: experienceSpecialFeatures = [],
+  } = experience;
+
+  const uniquePurposeIds = useMemo(() => {
```
Comment (allisonking) [inline clients/fides-js/src/components/tcf/InitialLayer.tsx:22]: we now have two fields: `tcf_consent_purposes` and `tcf_legitimate_interests_purposes` which can overlap! this means, if we only want one instance of a purpose, we need to dedupe

File: clients/fides-js/src/components/tcf/TcfOverlay.tsx [hunk 2]
```diff
@@ -212,63 +204,28 @@ const TcfOverlay: FunctionComponent<OverlayProps> = ({
...
+      tcf_special_purposes: specialPurposes = [],
+      tcf_features: features = [],
+      tcf_special_features: specialFeatures = [],
+      tcf_consent_vendors: consentVendors = [],
+      tcf_legitimate_interests_vendors: legintVendors = [],
+      tcf_consent_systems: consentSystems = [],
+      tcf_legitimate_interests_systems: legintSystems = [],
     } = experience;
 
-    const vendorsAndSystems = [...(vendors || []), ...(systems || [])];
```
Comment (allisonking) [inline clients/fides-js/src/components/tcf/TcfOverlay.tsx:223]: this was the bulk of the tech debt I removed 🧹

File: clients/fides-js/src/components/tcf/TcfPurposes.tsx [hunk 3]
```diff
@@ -114,49 +116,40 @@ const PurposeBlock = ({
...
-          onChange({ newEnabledIds, modelType: "specialPurposes" })
-        }
-        hideToggles
-      />
-    </div>
-  );
-};
+}) => (
+  <div>
+    <PurposeBlock
```
Comment (allisonking) [inline clients/fides-js/src/components/tcf/TcfPurposes.tsx:135]: Right now we are only rendering consent purposes—legint will follow in https://github.com/ethyca/fides/issues/4210

File: clients/fides-js/src/components/tcf/VendorInfoBanner.tsx [hunk 4]
```diff
@@ -37,27 +35,22 @@ const VendorInfoBanner = ({
...
-    const legint =
-      vendorRecordsWithLegalBasis(
-        systems,
-        LegalBasisForProcessingEnum.LEGITIMATE_INTERESTS
-      ).length +
-      vendorRecordsWithLegalBasis(
-        vendors,
-        LegalBasisForProcessingEnum.LEGITIMATE_INTERESTS
-      ).length;
+    const consent = consentSystems.length + consentVendors.length;
```
Comment (allisonking) [inline clients/fides-js/src/components/tcf/VendorInfoBanner.tsx:51]: counting also got much easier with the backend update 👍

File: clients/fides-js/src/lib/tcf/types.ts [hunk 5]
```diff
@@ -5,6 +5,16 @@ import type {
   UserConsentPreference,
 } from "../consent-types";
 
+export enum LegalBasisForProcessingEnum {
```
Comment (allisonking) [inline clients/fides-js/src/lib/tcf/types.ts:8]: these are mostly all copied from the updated admin-ui types. I rearranged them a bit in the file to make more intuitive sense (hopefully)

File: clients/fides-js/src/lib/tcf/vendors.ts [hunk 6]
```diff
@@ -10,19 +17,65 @@ export const vendorIsGvl = (
...
-    const hasApplicablePurposes = purposes?.filter((purpose) =>
-      purpose.legal_bases?.includes(legalBasis)
-    );
-    const hasApplicableSpecialPurposes = specialPurposes?.filter((purpose) =>
-      purpose.legal_bases?.includes(legalBasis)
-    );
-    return (
-      hasApplicablePurposes?.length || hasApplicableSpecialPurposes?.length
-    );
+const transformVendorDataToVendorRecords = ({
```
Comment (allisonking) [inline clients/fides-js/src/lib/tcf/vendors.ts:20]: this function is one of the larger changes in this PR. basically the vendor tab wants both vendors and systems together. on top of that, it also wants it combined across legal basis. this function returns a type that makes that easy

File: clients/fides-js/src/components/tcf/TcfVendors.tsx [hunk 7]
```diff
@@ -250,24 +252,14 @@ const TcfVendors = ({
...
-          vendorRecordsWithLegalBasis(
-            [vendor],
-            LegalBasisForProcessingEnum.CONSENT
-          ).length === 1;
-        const isLegint =
-          vendorRecordsWithLegalBasis(
-            [vendor],
-            LegalBasisForProcessingEnum.LEGITIMATE_INTERESTS
-          ).length === 1;
+          experience.gvl?.dataCategories;
```
Comment (eastandwestwind) [inline clients/fides-js/src/components/tcf/TcfVendors.tsx:262]: 🎆

**Review by eastandwestwind — APPROVED** [review]
Thanks for walking through this one with me on screenshare @allisonking !

**Comment by allisonking** [comment]
Updated a bunch of types after PR review changes to https://github.com/ethyca/fides/commit/8933d7d22f95f7007db9b54a1c4657f5a0167f53

**Pull request merged**